File: benchmarks/python/bench.py
"""Benchmark harness for the Python leg of the language shootout.

Writes results/python.json with wall-clock timings for each benchmark so
that `scripts/aggregate_results.py` can merge all four languages into a
single comparison CSV.
"""
from __future__ import annotations

import json
import pathlib
import time
from typing import Callable

import numpy as np

from ewma import ewma_vol
from mc_gpu import mc_terminal_prices
from ring_buffer import RingBuffer
from vwap import rolling_vwap

RESULTS_DIR = pathlib.Path(__file__).resolve().parents[2] / "results"


def _time_it(fn: Callable[[], None], iterations: int = 5) -> float:
    """Returns the best-of-`iterations` wall-clock time in seconds."""
    best = float("inf")
    for _ in range(iterations):
        start = time.perf_counter()
        fn()
        best = min(best, time.perf_counter() - start)
    return best


def bench_ring_buffer(n_push: int = 1_000_000) -> float:
    rb = RingBuffer(1 << 16)

    def run() -> None:
        for i in range(n_push):
            rb.push(float(i))

    return _time_it(run, iterations=3)


def bench_vwap(n: int = 1_000_000) -> float:
    rng = np.random.default_rng(0)
    px = 100.0 + rng.standard_normal(n).cumsum() * 0.01
    qty = rng.uniform(1.0, 100.0, size=n)

    def run() -> None:
        rolling_vwap(px, qty)

    return _time_it(run)


def bench_ewma(n: int = 1_000_000) -> float:
    rng = np.random.default_rng(0)
    ret = rng.standard_normal(n) * 0.01
    ewma_vol(ret, 0.94)  # Warm up the Numba JIT before timing.

    def run() -> None:
        ewma_vol(ret, 0.94)

    return _time_it(run)


def bench_mc(n_paths: int = 100_000, n_steps: int = 252) -> float:
    def run() -> None:
        mc_terminal_prices(100.0, 0.05, 0.20, 1.0 / 252.0, n_steps, n_paths)

    return _time_it(run, iterations=3)


def main() -> None:
    results = {
        "language": "python",
        "benchmarks": {
            "ring_buffer_push_1e6": bench_ring_buffer(),
            "vwap_1e6": bench_vwap(),
            "ewma_1e6": bench_ewma(),
            "monte_carlo_gbm_1e5x252": bench_mc(),
        },
    }
    RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    out_path = RESULTS_DIR / "python.json"
    out_path.write_text(json.dumps(results, indent=2))
    print(json.dumps(results, indent=2))


if __name__ == "__main__":
    main()

File: benchmarks/rust/src/bench.rs
//! Benchmark harness for the Rust leg of the language shootout.
//! Writes results/rust.json for scripts/aggregate_results.py to merge.

use std::fs;
use std::hint::black_box;
use std::path::PathBuf;
use std::time::Instant;

use rand::SeedableRng;
use rand_distr::{Distribution, Normal, Uniform};

use quant_lang_shootout::ewma::ewma_vol;
use quant_lang_shootout::mc_gpu::mc_terminal_prices;
use quant_lang_shootout::ring_buffer::RingBuffer;
use quant_lang_shootout::vwap::rolling_vwap;

fn time_it<F: FnMut()>(mut f: F, iterations: u32) -> f64 {
    let mut best = f64::INFINITY;
    for _ in 0..iterations {
        let start = Instant::now();
        f();
        let elapsed = start.elapsed().as_secs_f64();
        if elapsed < best {
            best = elapsed;
        }
    }
    best
}

fn bench_ring_buffer(n_push: usize) -> f64 {
    let mut rb: RingBuffer<f64, 65536> = RingBuffer::new();
    time_it(
        || {
            for i in 0..n_push {
                rb.push(black_box(i as f64));
            }
            black_box(&rb);
        },
        3,
    )
}

fn bench_vwap(n: usize) -> f64 {
    let mut rng = rand::rngs::StdRng::seed_from_u64(0);
    let normal = Normal::new(0.0_f64, 1.0).unwrap();
    let uniform = Uniform::new(1.0_f64, 100.0);

    let mut px = vec![0.0; n];
    let mut qty = vec![0.0; n];
    let mut level = 100.0;
    for i in 0..n {
        level += normal.sample(&mut rng) * 0.01;
        px[i] = level;
        qty[i] = uniform.sample(&mut rng);
    }
    let mut out = vec![0.0; n];
    time_it(
        || {
            rolling_vwap(&px, &qty, &mut out);
            black_box(&out);
        },
        5,
    )
}

fn bench_ewma(n: usize) -> f64 {
    let mut rng = rand::rngs::StdRng::seed_from_u64(0);
    let normal = Normal::new(0.0_f64, 0.01).unwrap();
    let ret: Vec<f64> = (0..n).map(|_| normal.sample(&mut rng)).collect();
    let mut out = vec![0.0; n];
    time_it(
        || {
            ewma_vol(&ret, 0.94, &mut out);
            black_box(&out);
        },
        5,
    )
}

fn bench_monte_carlo(n_paths: u32, n_steps: u32) -> f64 {
    time_it(
        || {
            let out = mc_terminal_prices(100.0, 0.05, 0.20, 1.0 / 252.0, n_steps, n_paths, 12345);
            black_box(&out);
        },
        3,
    )
}

fn main() {
    let ring_buffer_s = bench_ring_buffer(1_000_000);
    let vwap_s = bench_vwap(1_000_000);
    let ewma_s = bench_ewma(1_000_000);
    let mc_s = bench_monte_carlo(100_000, 252);

    let json = format!(
        "{{\n  \"language\": \"rust\",\n  \"benchmarks\": {{\n    \"ring_buffer_push_1e6\": {ring_buffer_s},\n    \"vwap_1e6\": {vwap_s},\n    \"ewma_1e6\": {ewma_s},\n    \"monte_carlo_gbm_1e5x252\": {mc_s}\n  }}\n}}\n"
    );

    let results_dir: PathBuf = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("results");
    fs::create_dir_all(&results_dir).expect("failed to create results directory");
    fs::write(results_dir.join("rust.json"), &json).expect("failed to write results/rust.json");

    println!(
        "ring_buffer_push_1e6={ring_buffer_s:.6} vwap_1e6={vwap_s:.6} ewma_1e6={ewma_s:.6} mc={mc_s:.6}"
    );
}
